pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarMetadata.java

```diff
@@ -0,0 +1,383 @@
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
+import static com.facebook.presto.spi.type.DateType.DATE;
+import static com.facebook.presto.spi.type.TimeType.TIME;
+import static com.facebook.presto.spi.type.TimestampType.TIMESTAMP;
+import static java.util.Objects.requireNonNull;
+
+public class PulsarMetadata implements ConnectorMetadata {
+
+    private final String connectorId;
+    private final PulsarAdmin pulsarAdmin;
+
+    private static final Logger log = Logger.get(PulsarMetadata.class);
+
+    @Inject
+    public PulsarMetadata(PulsarConnectorId connectorId, PulsarConnectorConfig pulsarConnectorConfig) {
+        this.connectorId = requireNonNull(connectorId, "connectorId is null").toString();
+        try {
+            this.pulsarAdmin = pulsarConnectorConfig.getPulsarAdmin();
+        } catch (PulsarClientException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public List<String> listSchemaNames(ConnectorSession session) {
+        List<String> prestoSchemas = new LinkedList<>();
+        try {
+            List<String> tenants = pulsarAdmin.tenants().getTenants();
+            for (String tenant : tenants) {
+                prestoSchemas.addAll(pulsarAdmin.namespaces().getNamespaces(tenant));
+            }
+        } catch (PulsarAdminException e) {
+            throw new RuntimeException("Failed to get schemas from pulsar", e);
+        }
+        return prestoSchemas;
+    }
+
+    @Override
+    public ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName) {
+        return new PulsarTableHandle(
+                this.connectorId,
+                tableName.getSchemaName(),
+                tableName.getTableName(),
+                tableName.getTableName());
+    }
+
+    @Override
+    public List<ConnectorTableLayoutResult> getTableLayouts(ConnectorSession session, ConnectorTableHandle table,
+                                                            Constraint<ColumnHandle> constraint,
+                                                            Optional<Set<ColumnHandle>> desiredColumns) {
+        PulsarTableHandle handle = convertTableHandle(table);
+        ConnectorTableLayout layout = new ConnectorTableLayout(new PulsarTableLayoutHandle(handle));
+        return ImmutableList.of(new ConnectorTableLayoutResult(layout, constraint.getSummary()));
+    }
+
+    @Override
+    public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTableLayoutHandle handle) {
+        return new ConnectorTableLayout(handle);
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle table) {
+        return getTableMetadata(convertTableHandle(table).toSchemaTableName(), true);
+    }
+
+    @Override
+    public List<SchemaTableName> listTables(ConnectorSession session, String schemaNameOrNull) {
```

Comment:
What happens if a schema contains tens of thousands of topics? Will it ever trash wrt mem?

Reply:
I don't think tens of thousands of strings will not use that much memory.